File: defmi/avalanche.py
"""Avalanche L1 connection for the DeFMI state machine.

The target is a dedicated Avalanche custom VM, not a generic EVM contract.  The
VM orders and validates asset registration, account opening and settlement
transactions.  This module is the production-facing JSON-RPC client and the
crash-safe projection bridge into :mod:`defmi.facility`.

Only opaque commitments and their authorization are sent.  Payment amounts,
prices, account owners and MPC shares are not RPC fields.
"""

from __future__ import annotations

import json
import ssl
import threading
import time
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Callable, Mapping

from .facility import (AccountOpening, AssetDefinition, DefmiFacility,
                       QuorumApproval, SettlementOrder, SettlementReceipt)


class AvalancheRpcError(RuntimeError):
    """An authenticated request reached the L1 endpoint but did not succeed."""


def _approval_json(approval: QuorumApproval) -> dict:
    return {
        "statement": approval.statement.hex(),
        "signerEpoch": approval.signer_epoch,
        "domain": approval.domain,
        "beforeRoot": approval.before_root.hex(),
        "approvals": [
            {"nodeID": item.node_id, "signature": item.signature.hex()}
            for item in approval.approvals
        ],
    }


def _asset_json(asset: AssetDefinition) -> dict:
    return {
        "assetID": asset.asset_id.hex(),
        "code": asset.code,
        "kind": asset.kind.value,
        "decimals": asset.decimals,
        "termsDigest": asset.terms_digest.hex(),
    }


def _opening_json(opening: AccountOpening) -> dict:
    return {
        "handle": opening.handle.hex(),
        "assetID": opening.asset_id.hex(),
        "commitment": opening.commitment.hex(),
        "issuanceNonce": opening.issuance_nonce.hex(),
    }


def _order_json(order: SettlementOrder) -> dict:
    return {
        "operationID": order.operation_id.hex(),
        "nullifier": order.nullifier.hex(),
        "deadline": order.deadline,
        "paymentInstructionDigest": order.payment_instruction_digest.hex(),
        "proofDigest": order.proof_digest.hex(),
        "marketStatementDigest": order.market_statement_digest.hex(),
        "legs": [
            {
                "handle": leg.handle.hex(),
                "assetID": leg.asset_id.hex(),
                "beforeCommitment": leg.before_commitment.hex(),
                "afterCommitment": leg.after_commitment.hex(),
                "beforeSequence": leg.before_sequence,
            }
            for leg in order.legs
        ],
    }


@dataclass(frozen=True)
class AcceptedTransition:
    tx_id: str
    block_id: str
    height: int
    statement: bytes
    before_root: bytes
    after_root: bytes

    @classmethod
    def parse(cls, value: Mapping) -> "AcceptedTransition":
        try:
            parsed = cls(
                tx_id=str(value["txID"]),
                block_id=str(value["blockID"]),
                height=int(value["height"]),
                statement=bytes.fromhex(value["statement"]),
                before_root=bytes.fromhex(value["beforeRoot"]),
                after_root=bytes.fromhex(value["afterRoot"]),
            )
        except (KeyError, TypeError, ValueError) as exc:
            raise AvalancheRpcError("L1 returned a malformed acceptance receipt") from exc
        if (
            parsed.height < 0
            or len(parsed.statement) != 32
            or len(parsed.before_root) != 32
            or len(parsed.after_root) != 32
        ):
            raise AvalancheRpcError("L1 acceptance receipt has invalid field widths")
        return parsed


class AvalancheRpcClient:
    """Small fail-closed client for the dedicated DeFMI VM JSON-RPC API."""

    def __init__(
        self,
        endpoint: str,
        *,
        timeout_seconds: float = 10.0,
        ssl_context: ssl.SSLContext | None = None,
        allow_insecure_localhost: bool = False,
        opener: Callable | None = None,
    ):
        parsed = urllib.parse.urlparse(endpoint)
        is_loopback = parsed.hostname in {"127.0.0.1", "localhost", "::1"}
        if parsed.scheme not in {"http", "https"} or not parsed.netloc:
            raise ValueError("Avalanche endpoint must be an absolute HTTP(S) URL")
        if parsed.scheme != "https" and not (allow_insecure_localhost and is_loopback):
            raise ValueError("plaintext Avalanche RPC is allowed only for an explicit localhost test")
        if timeout_seconds <= 0:
            raise ValueError("RPC timeout must be positive")
        self.endpoint = endpoint
        self.timeout_seconds = timeout_seconds
        self.ssl_context = ssl_context
        self._opener = opener or urllib.request.urlopen
        self._next_id = 1

    def call(self, method: str, params: Mapping | None = None):
        request_id = self._next_id
        self._next_id += 1
        body = json.dumps(
            {"jsonrpc": "2.0", "id": request_id, "method": method,
             "params": dict(params or {})},
            sort_keys=True,
            separators=(",", ":"),
        ).encode()
        request = urllib.request.Request(
            self.endpoint,
            data=body,
            method="POST",
            headers={"Content-Type": "application/json", "Accept": "application/json"},
        )
        try:
            with self._opener(
                request, timeout=self.timeout_seconds, context=self.ssl_context
            ) as response:
                raw = response.read(1_048_577)
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise AvalancheRpcError(f"Avalanche RPC transport failed: {exc}") from exc
        if len(raw) > 1_048_576:
            raise AvalancheRpcError("Avalanche RPC response exceeded one MiB")
        try:
            envelope = json.loads(raw)
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise AvalancheRpcError("Avalanche RPC returned invalid JSON") from exc
        if not isinstance(envelope, dict) or envelope.get("id") != request_id:
            raise AvalancheRpcError("Avalanche RPC response identifier does not match")
        if envelope.get("error") is not None:
            error = envelope["error"]
            message = error.get("message", "unknown error") if isinstance(error, dict) else str(error)
            raise AvalancheRpcError(f"Avalanche RPC rejected the request: {message}")
        if "result" not in envelope:
            raise AvalancheRpcError("Avalanche RPC response has no result")
        return envelope["result"]

    def state_root(self) -> bytes:
        result = self.call("defmivm.stateRoot")
        try:
            root = bytes.fromhex(result["stateRoot"])
        except (KeyError, TypeError, ValueError) as exc:
            raise AvalancheRpcError("L1 returned an invalid state root") from exc
        if len(root) != 32:
            raise AvalancheRpcError("L1 state root must be 32 bytes")
        return root

    def issue_asset(self, asset: AssetDefinition, approval: QuorumApproval,
                    expected_before_root: bytes) -> str:
        result = self.call(
            "defmivm.issueAsset", {"asset": _asset_json(asset),
                                   "approval": _approval_json(approval),
                                   "expectedBeforeRoot": expected_before_root.hex()}
        )
        return self._tx_id(result)

    def issue_account(self, opening: AccountOpening, approval: QuorumApproval,
                      expected_before_root: bytes) -> str:
        result = self.call(
            "defmivm.issueAccount", {"opening": _opening_json(opening),
                                     "approval": _approval_json(approval),
                                     "expectedBeforeRoot": expected_before_root.hex()}
        )
        return self._tx_id(result)

    def issue_settlement(self, order: SettlementOrder,
                         approval: QuorumApproval,
                         expected_before_root: bytes) -> str:
        result = self.call(
            "defmivm.issueSettlement", {"order": _order_json(order),
                                        "approval": _approval_json(approval),
                                        "expectedBeforeRoot": expected_before_root.hex()}
        )
        return self._tx_id(result)

    @staticmethod
    def _tx_id(result) -> str:
        if not isinstance(result, dict) or not isinstance(result.get("txID"), str):
            raise AvalancheRpcError("L1 did not return a transaction identifier")
        return result["txID"]

    def wait_accepted(self, tx_id: str, *, timeout_seconds: float = 30.0,
                      poll_seconds: float = 0.2) -> AcceptedTransition:
        if timeout_seconds <= 0 or poll_seconds <= 0:
            raise ValueError("acceptance timeout and polling interval must be positive")
        deadline = time.monotonic() + timeout_seconds
        while True:
            result = self.call("defmivm.txStatus", {"txID": tx_id})
            if not isinstance(result, dict):
                raise AvalancheRpcError("L1 returned a malformed transaction status")
            status = result.get("status")
            if status == "accepted":
                return AcceptedTransition.parse(result)
            if status == "rejected":
                raise AvalancheRpcError(
                    f"Avalanche consensus rejected transaction {tx_id}: "
                    f"{result.get('reason', 'unspecified')}"
                )
            # A node can briefly lose its local mempool view while a submitted
            # transaction is moving into consensus or while the node catches
            # up after a restart.  The returned transaction identifier is the
            # durable correlation key, so keep polling until acceptance,
            # explicit rejection or the caller's deadline.
            if status not in {"pending", "processing", "unknown"}:
                raise AvalancheRpcError(f"L1 returned unknown transaction status {status!r}")
            if time.monotonic() >= deadline:
                raise TimeoutError(f"Avalanche transaction {tx_id} was not accepted in time")
            time.sleep(poll_seconds)


class FacilityAvalancheBridge:
    """Keeps the durable local projection exactly aligned with Avalanche."""

    def __init__(self, facility: DefmiFacility, client: AvalancheRpcClient):
        self.facility = facility
        self.client = client
        self._lock = threading.RLock()

    def _require_aligned(self) -> bytes:
        local = self.facility.state_root()
        remote = self.client.state_root()
        if local != remote:
            raise RuntimeError(
                f"DeFMI projection is out of sync (local={local.hex()}, L1={remote.hex()})"
            )
        return local

    @staticmethod
    def _check(receipt: AcceptedTransition, statement: bytes,
               before_root: bytes) -> None:
        if receipt.statement != statement:
            raise RuntimeError("Avalanche accepted a different authorized statement")
        if receipt.before_root != before_root:
            raise RuntimeError("Avalanche applied the transition to an unexpected state root")

    def _require_approval(self, statement: bytes, before_root: bytes,
                          approval: QuorumApproval) -> None:
        if not self.facility.authorizer.verify(statement, before_root, approval):
            raise PermissionError(
                "the transition approval is not bound to this L1 and state root")

    def register_asset(self, asset: AssetDefinition,
                       approval: QuorumApproval) -> AcceptedTransition:
        with self._lock:
            before = self.facility.state_root()
            self._require_approval(asset.statement, before, approval)
            accepted = self.client.wait_accepted(
                self.client.issue_asset(asset, approval, before))
            self._check(accepted, asset.statement, before)
            self.facility.register_asset(asset, approval)
            if self.facility.state_root() != accepted.after_root:
                raise RuntimeError("asset projection root differs from Avalanche")
            return accepted

    def open_account(self, opening: AccountOpening,
                     approval: QuorumApproval) -> AcceptedTransition:
        with self._lock:
            before = self.facility.state_root()
            self._require_approval(opening.statement, before, approval)
            accepted = self.client.wait_accepted(
                self.client.issue_account(opening, approval, before))
            self._check(accepted, opening.statement, before)
            self.facility.open_account(opening, approval)
            if self.facility.state_root() != accepted.after_root:
                raise RuntimeError("account projection root differs from Avalanche")
            return accepted

    def settle(self, order: SettlementOrder, approval: QuorumApproval,
               *, now: int) -> tuple[SettlementReceipt, AcceptedTransition]:
        with self._lock:
            before = self.facility.state_root()
            self._require_approval(order.statement, before, approval)
            accepted = self.client.wait_accepted(
                self.client.issue_settlement(order, approval, before)
            )
            self._check(accepted, order.statement, before)
            local = self.facility.settle(order, approval, now=now)
            if (local.before_root != accepted.before_root
                    or local.after_root != accepted.after_root):
                raise RuntimeError("settlement projection root differs from Avalanche")
            return local, accepted

File: rust/qomm-defmi/src/avalanche.rs
//! Fail-closed Avalanche custom-VM JSON-RPC client and projection bridge.

use crate::facility::{
    AccountOpening, AssetDefinition, DefmiFacility, QuorumApproval, SettlementOrder,
    SettlementReceipt,
};
use openssl::ssl::{SslConnector, SslMethod, SslVerifyMode};
use serde_json::{json, Value};
use std::io::{Read, Write};
use std::net::TcpStream;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

const MAX_RESPONSE: usize = 1_048_576;

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AcceptedTransition {
    pub tx_id: String,
    pub block_id: String,
    pub height: u64,
    pub statement: [u8; 32],
    pub before_root: [u8; 32],
    pub after_root: [u8; 32],
}

impl AcceptedTransition {
    pub fn parse(value: &Value) -> Result<Self, String> {
        let object = value
            .as_object()
            .ok_or_else(|| "L1 returned a malformed acceptance receipt".to_string())?;
        let parse = |name: &str| -> Result<[u8; 32], String> {
            hex::decode(
                object
                    .get(name)
                    .and_then(Value::as_str)
                    .ok_or_else(|| "L1 returned a malformed acceptance receipt".to_string())?,
            )
            .map_err(|_| "L1 acceptance receipt has invalid field widths".to_string())?
            .try_into()
            .map_err(|_| "L1 acceptance receipt has invalid field widths".to_string())
        };
        Ok(Self {
            tx_id: object
                .get("txID")
                .and_then(Value::as_str)
                .ok_or_else(|| "L1 returned a malformed acceptance receipt".to_string())?
                .to_string(),
            block_id: object
                .get("blockID")
                .and_then(Value::as_str)
                .ok_or_else(|| "L1 returned a malformed acceptance receipt".to_string())?
                .to_string(),
            height: object
                .get("height")
                .and_then(Value::as_u64)
                .ok_or_else(|| "L1 returned a malformed acceptance receipt".to_string())?,
            statement: parse("statement")?,
            before_root: parse("beforeRoot")?,
            after_root: parse("afterRoot")?,
        })
    }
}

fn approval_json(approval: &QuorumApproval) -> Value {
    json!({
        "statement": hex::encode(approval.statement),
        "signerEpoch": approval.signer_epoch,
        "domain": approval.domain,
        "beforeRoot": hex::encode(approval.before_root),
        "approvals": approval.approvals.iter().map(|signed| json!({
            "nodeID": signed.node_id,
            "signature": hex::encode(signed.signature.to_bytes()),
        })).collect::<Vec<_>>(),
    })
}

fn asset_json(asset: &AssetDefinition) -> Value {
    json!({
        "assetID": hex::encode(asset.asset_id),
        "code": asset.code,
        "kind": asset.kind.as_str(),
        "decimals": asset.decimals,
        "termsDigest": hex::encode(asset.terms_digest),
    })
}

fn opening_json(opening: &AccountOpening) -> Value {
    json!({
        "handle": hex::encode(opening.handle),
        "assetID": hex::encode(opening.asset_id),
        "commitment": hex::encode(opening.commitment),
        "issuanceNonce": hex::encode(opening.issuance_nonce),
    })
}

fn order_json(order: &SettlementOrder) -> Value {
    json!({
        "operationID": hex::encode(order.operation_id),
        "nullifier": hex::encode(order.nullifier),
        "deadline": order.deadline,
        "paymentInstructionDigest": hex::encode(order.payment_instruction_digest),
        "proofDigest": hex::encode(order.proof_digest),
        "marketStatementDigest": hex::encode(order.market_statement_digest),
        "legs": order.legs.iter().map(|leg| json!({
            "handle": hex::encode(leg.handle),
            "assetID": hex::encode(leg.asset_id),
            "beforeCommitment": hex::encode(leg.before_commitment),
            "afterCommitment": hex::encode(leg.after_commitment),
            "beforeSequence": leg.before_sequence,
        })).collect::<Vec<_>>(),
    })
}

#[derive(Clone, Debug)]
struct Endpoint {
    tls: bool,
    host: String,
    port: u16,
    path: String,
    authority: String,
}

fn parse_endpoint(endpoint: &str, allow_insecure_localhost: bool) -> Result<Endpoint, String> {
    let (tls, rest) = if let Some(rest) = endpoint.strip_prefix("https://") {
        (true, rest)
    } else if let Some(rest) = endpoint.strip_prefix("http://") {
        (false, rest)
    } else {
        return Err("Avalanche endpoint must be an absolute HTTP(S) URL".into());
    };
    let (authority, path) = rest
        .split_once('/')
        .map_or((rest, "/".to_string()), |(authority, path)| {
            (authority, format!("/{path}"))
        });
    if authority.is_empty() || authority.contains('@') {
        return Err("Avalanche endpoint must be an absolute HTTP(S) URL".into());
    }
    let (host, port) = if let Some(bracketed) = authority.strip_prefix('[') {
        let (host, suffix) = bracketed
            .split_once(']')
            .ok_or_else(|| "Avalanche endpoint has an invalid IPv6 host".to_string())?;
        let port = suffix
            .strip_prefix(':')
            .map(|port| port.parse::<u16>())
            .transpose()
            .map_err(|_| "Avalanche endpoint has an invalid port".to_string())?
            .unwrap_or(if tls { 443 } else { 80 });
        (host.to_string(), port)
    } else {
        match authority.rsplit_once(':') {
            Some((host, port)) if !host.contains(':') => (
                host.to_string(),
                port.parse::<u16>()
                    .map_err(|_| "Avalanche endpoint has an invalid port".to_string())?,
            ),
            _ => (authority.to_string(), if tls { 443 } else { 80 }),
        }
    };
    if host.is_empty() {
        return Err("Avalanche endpoint must be an absolute HTTP(S) URL".into());
    }
    let loopback = matches!(host.as_str(), "127.0.0.1" | "localhost" | "::1");
    if !tls && !(allow_insecure_localhost && loopback) {
        return Err(
            "plaintext Avalanche RPC is allowed only for an explicit localhost test".into(),
        );
    }
    Ok(Endpoint {
        tls,
        host,
        port,
        path,
        authority: authority.to_string(),
    })
}

trait ReadWrite: Read + Write {}
impl<T: Read + Write> ReadWrite for T {}

fn decode_chunked(mut body: &[u8]) -> Result<Vec<u8>, String> {
    let mut decoded = Vec::new();
    loop {
        let line_end = body
            .windows(2)
            .position(|window| window == b"\r\n")
            .ok_or_else(|| "Avalanche RPC returned malformed chunked HTTP".to_string())?;
        let length = std::str::from_utf8(&body[..line_end])
            .ok()
            .and_then(|line| line.split(';').next())
            .and_then(|length| usize::from_str_radix(length.trim(), 16).ok())
            .ok_or_else(|| "Avalanche RPC returned malformed chunked HTTP".to_string())?;
        body = &body[line_end + 2..];
        if length == 0 {
            return Ok(decoded);
        }
        if body.len() < length + 2 || &body[length..length + 2] != b"\r\n" {
            return Err("Avalanche RPC returned malformed chunked HTTP".into());
        }
        decoded.extend_from_slice(&body[..length]);
        if decoded.len() > MAX_RESPONSE {
            return Err("Avalanche RPC response exceeded one MiB".into());
        }
        body = &body[length + 2..];
    }
}

fn http_post(endpoint: &Endpoint, body: &[u8], timeout: Duration) -> Result<Vec<u8>, String> {
    let tcp = TcpStream::connect((endpoint.host.as_str(), endpoint.port))
        .map_err(|error| format!("Avalanche RPC transport failed: {error}"))?;
    tcp.set_read_timeout(Some(timeout))
        .map_err(|error| error.to_string())?;
    tcp.set_write_timeout(Some(timeout))
        .map_err(|error| error.to_string())?;
    let mut stream: Box<dyn ReadWrite> = if endpoint.tls {
        let mut builder =
            SslConnector::builder(SslMethod::tls_client()).map_err(|error| error.to_string())?;
        builder.set_verify(SslVerifyMode::PEER);
        builder
            .set_default_verify_paths()
            .map_err(|error| error.to_string())?;
        Box::new(
            builder
                .build()
                .connect(&endpoint.host, tcp)
                .map_err(|error| format!("Avalanche RPC TLS authentication failed: {error}"))?,
        )
    } else {
        Box::new(tcp)
    };
    write!(
        stream,
        "POST {} HTTP/1.1\r\nHost: {}\r\nContent-Type: application/json\r\nAccept: application/json\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
        endpoint.path,
        endpoint.authority,
        body.len()
    )
    .map_err(|error| format!("Avalanche RPC transport failed: {error}"))?;
    stream
        .write_all(body)
        .map_err(|error| format!("Avalanche RPC transport failed: {error}"))?;
    stream
        .flush()
        .map_err(|error| format!("Avalanche RPC transport failed: {error}"))?;
    let mut raw = Vec::new();
    stream
        .take((MAX_RESPONSE + 65_536 + 1) as u64)
        .read_to_end(&mut raw)
        .map_err(|error| format!("Avalanche RPC transport failed: {error}"))?;
    let header_end = raw
        .windows(4)
        .position(|window| window == b"\r\n\r\n")
        .ok_or_else(|| "Avalanche RPC returned malformed HTTP".to_string())?;
    let headers = std::str::from_utf8(&raw[..header_end])
        .map_err(|_| "Avalanche RPC returned malformed HTTP headers".to_string())?;
    let mut lines = headers.split("\r\n");
    let status = lines
        .next()
        .and_then(|line| line.split_whitespace().nth(1))
        .and_then(|code| code.parse::<u16>().ok())
        .ok_or_else(|| "Avalanche RPC returned malformed HTTP status".to_string())?;
    if !(200..300).contains(&status) {
        return Err(format!(
            "Avalanche RPC HTTP endpoint returned status {status}"
        ));
    }
    let chunked = lines.any(|line| {
        line.split_once(':').is_some_and(|(name, value)| {
            name.eq_ignore_ascii_case("transfer-encoding")
                && value.to_ascii_lowercase().contains("chunked")
        })
    });
    let body = &raw[header_end + 4..];
    let body = if chunked {
        decode_chunked(body)?
    } else {
        body.to_vec()
    };
    if body.len() > MAX_RESPONSE {
        return Err("Avalanche RPC response exceeded one MiB".into());
    }
    Ok(body)
}

type Transport = dyn Fn(&[u8], Duration) -> Result<Vec<u8>, String> + Send + Sync;

pub struct AvalancheRpcClient {
    timeout: Duration,
    next_id: AtomicU64,
    transport: Arc<Transport>,
}

impl AvalancheRpcClient {
    pub fn new(
        endpoint: &str,
        timeout: Duration,
        allow_insecure_localhost: bool,
    ) -> Result<Self, String> {
        if timeout.is_zero() {
            return Err("RPC timeout must be positive".into());
        }
        let endpoint = parse_endpoint(endpoint, allow_insecure_localhost)?;
        Ok(Self {
            timeout,
            next_id: AtomicU64::new(1),
            transport: Arc::new(move |body, timeout| http_post(&endpoint, body, timeout)),
        })
    }

    pub fn with_transport<F>(
        endpoint: &str,
        timeout: Duration,
        allow_insecure_localhost: bool,
        transport: F,
    ) -> Result<Self, String>
    where
        F: Fn(&[u8], Duration) -> Result<Vec<u8>, String> + Send + Sync + 'static,
    {
        parse_endpoint(endpoint, allow_insecure_localhost)?;
        if timeout.is_zero() {
            return Err("RPC timeout must be positive".into());
        }
        Ok(Self {
            timeout,
            next_id: AtomicU64::new(1),
            transport: Arc::new(transport),
        })
    }

    pub fn call(&self, method: &str, params: Value) -> Result<Value, String> {
        let request_id = self.next_id.fetch_add(1, Ordering::Relaxed);
        let body = serde_json::to_vec(&json!({
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": params,
        }))
        .map_err(|error| error.to_string())?;
        let raw = (self.transport)(&body, self.timeout)?;
        if raw.len() > MAX_RESPONSE {
            return Err("Avalanche RPC response exceeded one MiB".into());
        }
        let envelope: Value = serde_json::from_slice(&raw)
            .map_err(|_| "Avalanche RPC returned invalid JSON".to_string())?;
        if envelope.get("id").and_then(Value::as_u64) != Some(request_id) {
            return Err("Avalanche RPC response identifier does not match".into());
        }
        if let Some(error) = envelope.get("error").filter(|value| !value.is_null()) {
            let message = error
                .get("message")
                .and_then(Value::as_str)
                .unwrap_or("unknown error");
            return Err(format!("Avalanche RPC rejected the request: {message}"));
        }
        envelope
            .get("result")
            .cloned()
            .ok_or_else(|| "Avalanche RPC response has no result".to_string())
    }

    fn transaction_id(result: &Value) -> Result<String, String> {
        result
            .get("txID")
            .and_then(Value::as_str)
            .map(str::to_string)
            .ok_or_else(|| "L1 did not return a transaction identifier".to_string())
    }
}

pub trait AvalancheClient: Send + Sync {
    fn state_root(&self) -> Result<[u8; 32], String>;
    fn issue_asset(
        &self,
        asset: &AssetDefinition,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String>;
    fn issue_account(
        &self,
        opening: &AccountOpening,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String>;
    fn issue_settlement(
        &self,
        order: &SettlementOrder,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String>;
    fn wait_accepted(
        &self,
        tx_id: &str,
        timeout: Duration,
        poll: Duration,
    ) -> Result<AcceptedTransition, String>;
}

impl AvalancheClient for AvalancheRpcClient {
    fn state_root(&self) -> Result<[u8; 32], String> {
        let result = self.call("defmivm.stateRoot", json!({}))?;
        let raw = result
            .get("stateRoot")
            .and_then(Value::as_str)
            .ok_or_else(|| "L1 returned an invalid state root".to_string())?;
        hex::decode(raw)
            .map_err(|_| "L1 returned an invalid state root".to_string())?
            .try_into()
            .map_err(|_| "L1 state root must be 32 bytes".to_string())
    }

    fn issue_asset(
        &self,
        asset: &AssetDefinition,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String> {
        Self::transaction_id(&self.call(
            "defmivm.issueAsset",
            json!({
                "asset": asset_json(asset),
                "approval": approval_json(approval),
                "expectedBeforeRoot": hex::encode(expected_before_root),
            }),
        )?)
    }

    fn issue_account(
        &self,
        opening: &AccountOpening,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String> {
        Self::transaction_id(&self.call(
            "defmivm.issueAccount",
            json!({
                "opening": opening_json(opening),
                "approval": approval_json(approval),
                "expectedBeforeRoot": hex::encode(expected_before_root),
            }),
        )?)
    }

    fn issue_settlement(
        &self,
        order: &SettlementOrder,
        approval: &QuorumApproval,
        expected_before_root: [u8; 32],
    ) -> Result<String, String> {
        Self::transaction_id(&self.call(
            "defmivm.issueSettlement",
            json!({
                "order": order_json(order),
                "approval": approval_json(approval),
                "expectedBeforeRoot": hex::encode(expected_before_root),
            }),
        )?)
    }

    fn wait_accepted(
        &self,
        tx_id: &str,
        timeout: Duration,
        poll: Duration,
    ) -> Result<AcceptedTransition, String> {
        if timeout.is_zero() || poll.is_zero() {
            return Err("acceptance timeout and polling interval must be positive".into());
        }
        let started = Instant::now();
        loop {
            let result = self.call("defmivm.txStatus", json!({"txID": tx_id}))?;
            let status = result.get("status").and_then(Value::as_str);
            match status {
                Some("accepted") => return AcceptedTransition::parse(&result),
                Some("rejected") => {
                    return Err(format!(
                        "Avalanche consensus rejected transaction {tx_id}: {}",
                        result
                            .get("reason")
                            .and_then(Value::as_str)
                            .unwrap_or("unspecified")
                    ));
                }
                Some("pending" | "processing" | "unknown") => {}
                other => return Err(format!("L1 returned unknown transaction status {other:?}")),
            }
            if started.elapsed() >= timeout {
                return Err(format!(
                    "Avalanche transaction {tx_id} was not accepted in time"
                ));
            }
            thread::sleep(poll);
        }
    }
}

pub struct FacilityAvalancheBridge<'a, C: AvalancheClient> {
    pub facility: &'a DefmiFacility,
    pub client: &'a C,
}

impl<'a, C: AvalancheClient> FacilityAvalancheBridge<'a, C> {
    pub const fn new(facility: &'a DefmiFacility, client: &'a C) -> Self {
        Self { facility, client }
    }

    fn require_approval(
        &self,
        statement: &[u8; 32],
        before: &[u8; 32],
        approval: &QuorumApproval,
    ) -> Result<(), String> {
        if self.facility.authorizer.verify(statement, before, approval) {
            Ok(())
        } else {
            Err("the transition approval is not bound to this L1 and state root".into())
        }
    }

    fn check(
        accepted: &AcceptedTransition,
        statement: &[u8; 32],
        before: &[u8; 32],
    ) -> Result<(), String> {
        if accepted.statement != *statement {
            return Err("Avalanche accepted a different authorized statement".into());
        }
        if accepted.before_root != *before {
            return Err("Avalanche applied the transition to an unexpected state root".into());
        }
        Ok(())
    }

    pub fn register_asset(
        &self,
        asset: &AssetDefinition,
        approval: &QuorumApproval,
    ) -> Result<AcceptedTransition, String> {
        let before = self.facility.state_root()?;
        let statement = asset.statement()?;
        self.require_approval(&statement, &before, approval)?;
        let transaction = self.client.issue_asset(asset, approval, before)?;
        let accepted = self.client.wait_accepted(
            &transaction,
            Duration::from_secs(30),
            Duration::from_millis(200),
        )?;
        Self::check(&accepted, &statement, &before)?;
        self.facility.register_asset(asset, approval)?;
        if self.facility.state_root()? != accepted.after_root {
            return Err("asset projection root differs from Avalanche".into());
        }
        Ok(accepted)
    }

    pub fn open_account(
        &self,
        opening: &AccountOpening,
        approval: &QuorumApproval,
    ) -> Result<AcceptedTransition, String> {
        let before = self.facility.state_root()?;
        let statement = opening.statement()?;
        self.require_approval(&statement, &before, approval)?;
        let transaction = self.client.issue_account(opening, approval, before)?;
        let accepted = self.client.wait_accepted(
            &transaction,
            Duration::from_secs(30),
            Duration::from_millis(200),
        )?;
        Self::check(&accepted, &statement, &before)?;
        self.facility.open_account(opening, approval)?;
        if self.facility.state_root()? != accepted.after_root {
            return Err("account projection root differs from Avalanche".into());
        }
        Ok(accepted)
    }

    pub fn settle(
        &self,
        order: &SettlementOrder,
        approval: &QuorumApproval,
        now: u64,
    ) -> Result<(SettlementReceipt, AcceptedTransition), String> {
        let before = self.facility.state_root()?;
        let statement = order.statement()?;
        self.require_approval(&statement, &before, approval)?;
        let transaction = self.client.issue_settlement(order, approval, before)?;
        let accepted = self.client.wait_accepted(
            &transaction,
            Duration::from_secs(30),
            Duration::from_millis(200),
        )?;
        Self::check(&accepted, &statement, &before)?;
        let receipt = self.facility.settle(order, approval, now)?;
        if receipt.before_root != accepted.before_root || receipt.after_root != accepted.after_root
        {
            return Err("settlement projection root differs from Avalanche".into());
        }
        Ok((receipt, accepted))
    }
}
